cpp/include/cudf/detail/utilities/rmm_host_vector.hpp

```diff
@@ -0,0 +1,207 @@
+ *  \see https://en.cppreference.com/w/cpp/memory/allocator
+ */
+template <typename T>
+class rmm_host_allocator;
+
+/*! \p rmm_host_allocator is a CUDA-specific host memory allocator
+ *  that employs \c an `cudf::host_async_resource_ref` for allocation.
+ *
+ * This implementation is ported from pinned_host_vector in cudf.
+ *
+ *  \see https://en.cppreference.com/w/cpp/memory/allocator
+ */
+template <>
+class rmm_host_allocator<void> {
+ public:
+  using value_type      = void;            ///< The type of the elements in the allocator
+  using pointer         = void*;           ///< The type returned by address() / allocate()
+  using const_pointer   = void const*;     ///< The type returned by address()
+  using size_type       = std::size_t;     ///< The type used for the size of the allocation
+  using difference_type = std::ptrdiff_t;  ///< The type of the distance between two pointers
+
+  /**
+   * @brief converts a `rmm_host_allocator<void>` to `rmm_host_allocator<U>`
+   */
+  template <typename U>
+  struct rebind {
+    using other = rmm_host_allocator<U>;  ///< The rebound type
+  };
+};
+
+/*! \p rmm_host_allocator is a CUDA-specific host memory allocator
+ *  that employs \c `rmm::host_async_resource_ref` for allocation.
+ *
+ * This implementation is ported from pinned_host_vector in cudf.
+ *
+ *  \see https://en.cppreference.com/w/cpp/memory/allocator
+ */
+template <typename T>
+class rmm_host_allocator {
+ public:
+  using value_type      = T;               ///< The type of the elements in the allocator
+  using pointer         = T*;              ///< The type returned by address() / allocate()
+  using const_pointer   = T const*;        ///< The type returned by address()
+  using reference       = T&;              ///< The parameter type for address()
+  using const_reference = T const&;        ///< The parameter type for address()
+  using size_type       = std::size_t;     ///< The type used for the size of the allocation
+  using difference_type = std::ptrdiff_t;  ///< The type of the distance between two pointers
+
+  typedef cuda::std::true_type propagate_on_container_move_assignment;
+
+  /**
+   * @brief converts a `rmm_host_allocator<T>` to `rmm_host_allocator<U>`
+   */
+  template <typename U>
+  struct rebind {
+    using other = rmm_host_allocator<U>;  ///< The rebound type
+  };
+
+  /**
+   * @brief Cannot declare an empty host allocator.
+   */
+  rmm_host_allocator() = delete;
+
+  /**
+   * @brief Construct from a `cudf::host_async_resource_ref`
+   */
+  rmm_host_allocator(rmm::host_async_resource_ref _mr, rmm::cuda_stream_view _stream)
+    : mr(_mr), stream(_stream)
+  {
+  }
+
+  /**
+   * @brief Copy constructor
+   */
+  rmm_host_allocator(rmm_host_allocator const& other) = default;
+
+  /**
+   * @brief Move constructor
+   */
+  rmm_host_allocator(rmm_host_allocator&& other) = default;
+
+  /**
+   * @brief Assignment operator
+   */
+  rmm_host_allocator& operator=(rmm_host_allocator const& col)
```

Comment:
Nit, non-blocking. I wouldn't call this `col`.

```suggestion
  rmm_host_allocator& operator=(rmm_host_allocator const& other)
```